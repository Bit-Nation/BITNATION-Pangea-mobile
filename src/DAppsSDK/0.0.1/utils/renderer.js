// @flow

import React from 'react';
import components from '../components';

export type ComponentsJSON = { type: string, props: Object, children: Array<Object> };

/**
 * @desc Function to get element class or function from text representation, e.g. 'Text' means Text component.
 * @param {string} typeName String representation of the type.
 * @return {*} Class or function of component or null if type is unknown.
 */
export const getTypeElementFromText = (typeName: string) => components[typeName];

/**
 * @desc Validates props against list of valid names and filter only whitelisted ones.
 * Used to prevent DApp developer to pass something bad to components we provide.
 * @param {Object} props Props to validate.
 * @param {string[]} validProps Object containing arrays by keys `native`, `stateBased` and `callbacks` according to corresponding type of props.
 * @param {string} type String representation of component type.
 * @return {Object} Props filtered by groups.
 */
export const validateProps = (
  props: Object,
  validProps: { native: Array<string>, custom: Array<string>, callbacks: Array<string> },
  type: string,
) => {
  const nativeProps = {};
  const customProps = {};
  const callbackProps = {};
  Object.keys(props).forEach((propName) => {
    if (typeof propName !== 'string') {
      console.warn('Prop name is not a string');
      return;
    }

    if (validProps.native.includes(propName)) {
      // It's an allowed native prop.
      nativeProps[propName] = props[propName];
    } else if (validProps.custom.includes(propName)) {
      // It's an allowed custom prop.
      customProps[propName] = props[propName];
    } else if (propName.endsWith('ID') && validProps.callbacks.includes(propName.substring(0, propName.length - 2))) {
      // It's an allowed prop to perform a callback. That properties are marked by ID suffix.
      callbackProps[propName] = props[propName];
    } else {
      console.warn(`Prop ${propName} is not allowed on component ${type}`);
    }
  });

  return { nativeProps, customProps, callbackProps };
};

/**
 * @desc Renders JSON provided by DApp developer into react components.
 * @param {ComponentsJSON} json JSON object that represents components to render.
 * @param {?string} key Key of component (used for updates).
 * @param {Function} customPropsProvider Function that takes component class or function and props and returns new props if it's needed to add ones.
 * @param {*} parent Function or class of component that is parent of currently used or undefined for root.
 * @return {*} Tree of components that is ready to be render.
 */
export const renderJSON = (json: ComponentsJSON, key: ?string, customPropsProvider: (component: any, ownProps: Object) => Object, parent: any) => {
  const { type, children } = json;

  // This is the case when we have string literal inside children.
  // It's only allowed for Text.
  if (type === undefined && typeof (json) === 'string') {
    if (parent === components.Text) {
      return json;
    }

    console.warn(`String literal may only appear inside Text component. Found "${json}" outside.`);
    return null;
  }

  const component = components[type];

  if (component == null) {
    console.warn(`Trying to render unknown component type ${type}.`);
    return null;
  }

  if (component.validNativeProps != null && Array.isArray(component.validNativeProps) === false) {
    console.warn(`validNativeProps field of component ${type} should be an array of strings.`);
    return null;
  }
  if (component.customProps != null && Array.isArray(component.customProps) === false) {
    console.warn(`customProps field of component ${type} should be an array of strings.`);
    return null;
  }
  if (component.callbackProps != null && Array.isArray(component.callbackProps) === false) {
    console.warn(`callbackProps field of component ${type} should be an array of strings.`);
    return null;
  }

  const { props } = json;

  const { nativeProps = props, customProps = {}, callbackProps = {} } =
    validateProps(json.props, {
      native: component.validNativeProps || [],
      custom: component.customProps || [],
      callbacks: component.callbackProps || [],
    }, type);

  return React.createElement(
    component,
    {
      nativeProps,
      ...customProps,
      ...customPropsProvider(component, { ...nativeProps, ...customProps, ...callbackProps }),
      key,
    },
    children
      ? children.map((child, index) => renderJSON(child, `${index}`, customPropsProvider, component))
      : null,
  );
};
