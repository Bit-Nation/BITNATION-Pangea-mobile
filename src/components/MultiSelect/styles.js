/*!
 * react-native-multi-select
 * Copyright(c) 2017 Mustapha Babatunde Oluwaleke
 * MIT Licensed
 */
import Colors from '../../global/colors';

export const colorPack = {
  primary: Colors.BitnationHighlightColor,
  primaryDark: '#215191',
  light: 'transparent', // Colors.shadeOfBitnationLightColor(0.2),
  textPrimary: Colors.BitnationDarkGrayColor,
  placeholderTextColor: Colors.BitnationDarkGrayColor,
  danger: Colors.primary_red,
  borderColor: Colors.borderColor,
  backgroundColor: 'transparent',
};

export default {
  footerWrapper: {
    flexWrap: 'wrap',
    alignItems: 'flex-start',
    flexDirection: 'row',
  },
  footerWrapperNC: {
    width: 320,
    flexDirection: 'column',
  },
  subSection: {
    backgroundColor: 'transparent',
    borderBottomWidth: 1,
    borderColor: colorPack.borderColor,
    paddingLeft: 0,
    paddingRight: 20,
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
  },
  greyButton: {
    height: 40,
    borderRadius: 5,
    elevation: 0,
    backgroundColor: colorPack.actionButtonColor,
  },
  indicator: {
    fontSize: 15,
    color: colorPack.placeholderTextColor,
  },
  selectedItem: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingLeft: 15,
    paddingTop: 3,
    paddingRight: 3,
    paddingBottom: 3,
    margin: 3,
    borderRadius: 20,
    borderWidth: 2,
  },
  button: {
    height: 25,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
  },
  buttonText: {
    color: Colors.white,
    fontSize: 16,
  },
  selectorView: (fixedHeight) => {
    const style = {
      flexDirection: 'column',
      marginBottom: 0,
      elevation: 2,
    };
    if (fixedHeight) {
      style.height = 250;
    }
    return style;
  },
  inputGroup: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingLeft: 6,
    backgroundColor: 'transparent',
  },
  dropdownView: {
    flexDirection: 'row',
    alignItems: 'center',
    height: 43,
    marginBottom: 0,
  },
};
