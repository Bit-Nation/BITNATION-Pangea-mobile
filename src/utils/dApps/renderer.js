// @flow

import React from 'react';

import Text from '../../components/dApps/Text';
import View from '../../components/dApps/View';
import TextInput from '../../components/dApps/TextInput';
import Button from '../../components/dApps/Button';

export type ComponentsJSON = { type: string, props: Object, children: Array<Object> };

/**
 * @desc Function to get element class or function from text representation, e.g. 'Text' means Text component.
 * @param {string} typeName String representation of the type.
 * @return {*} Class or function of component or null if type is unknown.
 */
export const getTypeElementFromText = (typeName: string) => {
  switch (typeName) {
    case 'Text':
      return Text;
    case 'View':
      return View;
    case 'TextInput':
      return TextInput;
    case 'Button':
      return Button;
    default:
      console.warn(`Unknown component type ${typeName}.`);
      return null;
  }
};

/**
 * @desc Validates props against list of valid names and filter only whitelisted ones.
 * Used to prevent DApp developer to pass something bad to components we provide.
 * @param {Object} props Props to validate.
 * @param {string[]} validProps Object containing arrays by keys `native`, `stateBased` and `callbacks` according to corresponding type of props.
 * @param {string} type String representation of component type.
 * @return {Object} Props filtered by groups.
 */
export const validateProps = (
  props: Object,
  validProps: { native: Array<string>, stateBased: Array<string>, callbacks: Array<string> },
  type: string,
) => {
  const nativeProps = {};
  const stateBasedProps = {};
  const callbackProps = {};
  Object.keys(props).forEach((propName) => {
    if (validProps.native.includes(propName)) {
      // It's an allowed native prop.
      nativeProps[propName] = props[propName];
    } else if (propName.endsWith('Path') && validProps.stateBased.includes(propName.substring(0, propName.length - 4))) {
      // It's an allowed prop to set or get root component state. That properties are marked by Path suffix.
      stateBasedProps[propName] = props[propName];
    } else if (propName.endsWith('ID') && validProps.callbacks.includes(propName.substring(0, propName.length - 2))) {
      // It's an allowed prop to perform a callback. That properties are marked by ID suffix.
      callbackProps[propName] = props[propName];
    } else {
      console.warn(`Prop ${propName} is not allowed on component ${type}`);
    }
  });

  return { nativeProps, stateBasedProps, callbackProps };
};

/**
 * @desc Renders JSON provided by DApp developer into react components.
 * @param {ComponentsJSON} json JSON object that represents components to render.
 * @param {?string} key Key of component (used for updates).
 * @param {Function} customPropsProvider Function that takes component class or function and props and returns new props if it's needed to add ones.
 * @param {*} parent Function or class of component that is parent of currently used or undefined for root.
 * @return {*} Tree of components that is ready to be render.
 */
export const renderJSON = (json: ComponentsJSON, key: ?string, customPropsProvider: (component: any, ownProps: Object) => Object, parent: any) => {
  const { type, children } = json;

  // This is the case when we have string literal inside children.
  // It's only allowed for Text.
  if (type === undefined && typeof (json) === 'string') {
    if (parent === Text) {
      return json;
    }

    console.warn(`String literal may only appear inside Text component. Found "${json}" outside.`);
    return null;
  }

  const component = getTypeElementFromText(type);

  if (component === null) {
    return null;
  }

  const { props } = json;

  const { nativeProps = props, stateBasedProps = {}, callbackProps = {} } =
    component.validNativeProps !== undefined
      ? validateProps(json.props, {
        native: component.validNativeProps,
        stateBased: Object.keys(component.stateBasedProps || {}),
        callbacks: component.callbackProps || [],
      }, type)
      : {};

  return React.createElement(
    component,
    { ...nativeProps, ...customPropsProvider(component, { ...nativeProps, ...stateBasedProps, ...callbackProps }), key },
    children
      ? children.map((child, index) => renderJSON(child, `${index}`, customPropsProvider, component))
      : null,
  );
};
