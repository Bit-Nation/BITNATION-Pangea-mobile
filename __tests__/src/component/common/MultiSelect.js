import MultiSelect from '../../../../src/components/MultiSelect';
import React from 'react';
import renderer from 'react-test-renderer'
import Colors from "../../../../src/global/Colors";

test('MultiSelect renders correctly', () => {
  const tree = renderer.create(
    <MultiSelect
      hideTags
      items={[{
        id: 'id1',
        name: 'Testing Opt 1',
      }, {
        id: 'id2',
        name: 'Testing Opt 2',
      }]}
      uniqueKey='id'
      submitButtonText={'OK'}
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test('MultiSelect renders correctly with customized colors', () => {
  const tree = renderer.create(
    <MultiSelect
      hideTags
      items={[{
        id: 'id1',
        name: 'Testing Opt 1',
      }, {
        id: 'id2',
        name: 'Testing Opt 2',
      }]}
      uniqueKey='id'
      tagRemoveIconColor='#CCC'
      tagBorderColor='#CCC'
      tagTextColor='#CCC'
      itemTextColor={Colors.textSecondary}
      displayKey='name'
      submitButtonColor={Colors.panelBoxColor}
      submitButtonText={'OK'}
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});