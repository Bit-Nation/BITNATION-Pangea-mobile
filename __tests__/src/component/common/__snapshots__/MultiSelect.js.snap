// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelect renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "flexDirection": "column",
      "marginBottom": 10,
      "marginTop": 10,
    }
  }
>
  <View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 40,
          "marginBottom": 10,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "rgba(192,192,192,0.2)",
              "borderBottomWidth": 1,
              "borderColor": "rgba(192,192,192,0.4)",
              "flex": 1,
              "flexDirection": "row",
              "paddingLeft": 20,
              "paddingRight": 20,
            },
            Object {
              "paddingBottom": 10,
              "paddingTop": 10,
            },
          ]
        }
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hitSlop={undefined}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
            }
          }
          testID={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            numberOfLines={1}
            style={
              Array [
                Object {
                  "color": "white",
                  "flex": 1,
                  "fontSize": 17,
                },
                Object {},
              ]
            }
          >
            Select
          </Text>
          <Text
            accessible={true}
            allowFontScaling={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 12,
                },
                Object {
                  "color": "white",
                  "fontSize": 30,
                },
                Object {
                  "fontFamily": "Ionicons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
              ]
            }
          >
            
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`MultiSelect renders correctly with customized colors 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "flexDirection": "column",
      "marginBottom": 10,
      "marginTop": 10,
    }
  }
>
  <View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 40,
          "marginBottom": 10,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "rgba(192,192,192,0.2)",
              "borderBottomWidth": 1,
              "borderColor": "rgba(192,192,192,0.4)",
              "flex": 1,
              "flexDirection": "row",
              "paddingLeft": 20,
              "paddingRight": 20,
            },
            Object {
              "paddingBottom": 10,
              "paddingTop": 10,
            },
          ]
        }
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hitSlop={undefined}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
            }
          }
          testID={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            numberOfLines={1}
            style={
              Array [
                Object {
                  "color": "white",
                  "flex": 1,
                  "fontSize": 17,
                },
                Object {},
              ]
            }
          >
            Select
          </Text>
          <Text
            accessible={true}
            allowFontScaling={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 12,
                },
                Object {
                  "color": "white",
                  "fontSize": 30,
                },
                Object {
                  "fontFamily": "Ionicons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
              ]
            }
          >
            
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
